webview/src/shared/components/contextMenu/ContextMenu.tsx

```diff
@@ -1,47 +1,45 @@
-import React from 'react'
+import React, { Dispatch, SetStateAction } from 'react'
 import { Instance } from 'tippy.js'
 import Tooltip from '../tooltip/Tooltip'
 
-const positionContextMenuAndDisableEvents = (
-  instance: Instance,
-  event: PointerEvent
-) => {
-  event.preventDefault()
-  instance.setProps({
-    getReferenceClientRect() {
-      const { top, bottom, height } = instance.reference.getBoundingClientRect()
-      return {
-        bottom,
-        height,
-        left: event.clientX,
-        right: event.clientX,
-        top,
-        width: 0
-      } as DOMRect
+const positionContextMenuAndDisableEvents =
+  (setHideOnClick: Dispatch<SetStateAction<(() => void) | undefined>>) =>
+  (instance: Instance, event: PointerEvent) => {
+    event.preventDefault()
+    instance.setProps({
+      getReferenceClientRect() {
+        const { top, bottom, height } =
+          instance.reference.getBoundingClientRect()
+        return {
+          bottom,
+          height,
+          left: event.clientX,
+          right: event.clientX,
+          top,
+          width: 0
+        } as DOMRect
+      }
+    })
+
+    const handleDefaultEvent = (e: Event) => {
+      e.preventDefault()
     }
-  })
 
-  const handleDefaultEvent = (e: Event) => {
-    e.preventDefault()
-  }
+    instance.reference.removeEventListener('contextmenu', handleDefaultEvent)
+    instance.reference.addEventListener('contextmenu', handleDefaultEvent)
 
-  instance.reference.removeEventListener('contextmenu', handleDefaultEvent)
-  instance.reference.addEventListener('contextmenu', handleDefaultEvent)
+    const hideOnClick = () => {
+      !instance.state.isDestroyed && instance.hide()
+    }
 
-  const hideOnClick = () => {
-    !instance.state.isDestroyed && instance.hide()
+    setHideOnClick(() => hideOnClick)
```

Comment:
See [also](https://github.com/iterative/vscode-dvc/pull/3825/files#r1184809441)

Reply:
How about passing a prop `shouldHide` to the `ContextMenu` and hiding inside a `useEffect`? Bringing the function upward feels backward  for some reason. This might also cause re-rendering problems as arrow functions are re-created on renders and passing it back up to be added to a state will trigger a re-render...

Reply:
Would a `clickOutiside` check also work here? Reducing the number of places that need to know about closing the context menu?

Reply:
Is the only reason we're needing this `setHideOnClick` prop is to make sure clicking on disabled items doesn't close the context menu?  If so, why use a window event listener like before but just check the element that was clicked on? Aka similar to what we had before:

```
  const hideOnClick = e => {
    if (!e.target.closest('[aria-disabled=true]')) {
      !instance.state.isDestroyed && instance.hide()
    }
  }

  instance.popper.removeEventListener('click', hideOnClick)
  instance.popper.addEventListener('click', hideOnClick)

window.addEventListener('click', hideOnClick, { once: true })
```

Seems to working from a quick test on my end. For the record, as part of this test, I commented out our `const [hideOnClick, setHideOnClick] = useState...` declarations and added a `aria-disabled` attribute to our disabled menu options:

https://user-images.githubusercontent.com/43496356/236240085-a4d998ce-86c8-46b7-967a-8bc77afefde0.mov